### DISCUSSION TITLE
usedExports optimization doesn't work when devtool is set to eval-source-map

### DISCUSSION THREAD

**Discussion by dmt0** [webpack/webpack#16736] [Q&A]
# Bug report

Asked a question on SO and later found the answer myself:
https://stackoverflow.com/questions/75077181/cant-get-webpack-usedexport-treeshaking-to-work/75087281#75087281

**What is the current behavior?**
When `devtool` is set to `eval-source-map`, `usedExports` settings has no effect.

**If the current behavior is a bug, please provide the steps to reproduce.**
Set `devtool` to `eval-source-map`, set `usedExports` to `true`, observe that the optimization does not happen.
Repro here: https://github.com/dmt0/treeshake-repro

**What is the expected behavior?**
I'm not exactly sure if this is itself a bug. Perhaps setting `devtool` to a certain value makes `webpack` assume that it's not in `production` mode in which case `usedExports` doesn't make sense. In any case it's not very clear from the docs and it took me a day of debugging until I've tried everything and started removing lines one by one from `webpack.config.js`. If this is indeed a desired behavior, perhaps there needs to be a mention somewhere in the docs.

Treeshaking docs (https://webpack.js.org/guides/tree-shaking/#conclusion) mention all the conditions necessary for it to work. Again `devtool` is not mentioned.

**Other relevant information:**
webpack version:  5.75.0
Node.js version: 16.14.0
Operating System: macOS
Additional tools: "@babel/core": "7.20.12"

**Comment by alexander-akait** [comment]
Yes, that is expected, because `eval` can't allow to analyze code, you should not use `eval-source-map` for production builds, `usedExports` is used only for production, why you need `eval-source-map` in such case?

**Reply by dmt0** [reply]
Hi!

In my case `eval-source-map` was used by mistake. My point is that it took me forever to diagnose why `usedExports` did not work. It might be obvious for you that `devtool` setting can impact `usedExports`, but it definitely wasn't obvious for me, a simple webpack user. There was no diagnostic message telling me that I'm doing something weird. It was not clear from any of the documentation that I linked in my original comment. There are many conditions stated there that are required for `usedExports` to do its job, but correct `devtool` setting is not one of them.

Perhaps this issue is meant as a documentation improvement issue.
